src/lens.js

```diff
@@ -1,51 +1,10 @@
-var _curry2 = require('./internal/_curry2');
+var curry = require('./curry');
+var map = require('./map');
 
 
 /**
- * Creates a lens. Supply a function to `get` values from inside an object, and a `set`
- * function to change values on an object. (n.b.: This can, and should, be done without
- * mutating the original object!) The lens is a function wrapped around the input `get`
- * function, with the `set` function attached as a property on the wrapper. A `map`
- * function is also attached to the returned function that takes a function to operate
- * on the specified (`get`) property, which is then `set` before returning. The attached
- * `set` and `map` functions are curried.
- *
- * @func
- * @memberOf R
- * @category Object
- * @sig (k -> v) -> (v -> a -> *) -> (a -> b)
- * @param {Function} get A function that gets a value by property name
- * @param {Function} set A function that sets a value by property name
- * @return {Function} the returned function has `set` and `map` properties that are
- *         also curried functions.
- * @example
- *
- *      var headLens = R.lens(
- *        function get(arr) { return arr[0]; },
- *        function set(val, arr) { return [val].concat(arr.slice(1)); }
- *      );
- *      headLens([10, 20, 30, 40]); //=> 10
- *      headLens.set('mu', [10, 20, 30, 40]); //=> ['mu', 20, 30, 40]
- *      headLens.map(function(x) { return x + 1; }, [10, 20, 30, 40]); //=> [11, 20, 30, 40]
- *
- *      var phraseLens = R.lens(
- *        function get(obj) { return obj.phrase; },
- *        function set(val, obj) {
- *          var out = R.clone(obj);
- *          out.phrase = val;
- *          return out;
- *        }
- *      );
- *      var obj1 = { phrase: 'Absolute filth . . . and I LOVED it!'};
- *      var obj2 = { phrase: "What's all this, then?"};
- *      phraseLens(obj1); // => 'Absolute filth . . . and I LOVED it!'
- *      phraseLens(obj2); // => "What's all this, then?"
- *      phraseLens.set('Ooh Betty', obj1); //=> { phrase: 'Ooh Betty'}
- *      phraseLens.map(R.toUpper, obj2); //=> { phrase: "WHAT'S ALL THIS, THEN?"}
+ * @sig $0 -> $1 -> $2 -> $3 -> $4
```

Comment:
`f` must return a functor, since we `map` over the result:

    Functor f => ($0 -> $1) -> ($2,$0 -> $3) -> ($1 -> f $4) -> $5 -> $6

If we perform substitutions in `map :: (a -> b) -> f a -> f b` for our application of `map` we get:

    ($4 -> $3) -> f $4 -> f $3

So, we can replace `$6` with `f $3`:

    Functor f => ($0 -> $1) -> ($2,$0 -> $3) -> ($1 -> f $4) -> $5 -> f $3

Also, since `setter :: ($2,$0 -> $3)` and the first argument to `setter` is a `$4` (from `($4 -> $3) -> f $4 -> f $3` above), `$2` and `$4` are the same type:

    Functor f => ($0 -> $1) -> ($2,$0 -> $3) -> ($1 -> f $2) -> $4 -> f $3

The last argument to `lens` is now a `$4`. This is given as the first argument to `getter` (`$0`) and as the second argument to `setter` (`$0`). So, `$0` and `$4` are the same type:

    Functor f => ($0 -> $1) -> ($2,$0 -> $3) -> ($1 -> f $2) -> $0 -> f $3

Substituting the identifiers gives:

    Functor f => (a -> b) -> (s,a -> t) -> (b -> f s) -> a -> f t

Is this reasoning sound?